pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -277,12 +281,36 @@ func (b *volumeBinder) BindPodVolumes(assumedPod *v1.Pod) error {
 		return err
 	}
 
-	return wait.Poll(time.Second, b.bindTimeout, func() (bool, error) {
+	bid, eventCh := b.broadcaster.Register()
+	defer b.broadcaster.Unregister(bid)
+	glog.V(5).Infof("got listener id %v", bid)
+
+	timer := time.NewTimer(b.bindTimeout)
+	defer timer.Stop()
+
+	for {
+		glog.V(5).Infof("BindPodVolumes: starting checkBindings for pod %q", podName)
+
 		// Get cached values every time in case the pod gets deleted
 		bindings = b.podBindingCache.GetBindings(assumedPod, assumedPod.Spec.NodeName)
 		claimsToProvision = b.podBindingCache.GetProvisionedPVCs(assumedPod, assumedPod.Spec.NodeName)
-		return b.checkBindings(assumedPod, bindings, claimsToProvision)
-	})
+		allBound, err := b.checkBindings(assumedPod, bindings, claimsToProvision)
+		if err != nil {
+			return err
+		}
+		if allBound {
+			glog.V(4).Infof("BindPodVolumes: all volumes bound for pod %q", podName)
+			return nil
+		}
+
+		// Wait for API updates
+		select {
+		case <-eventCh:
```

Comment:
As we cannot tell which events are indeed related to a certain pod, what worries me most is that in production environment, there might be quite a lot of events, and `checkBindings` will be called more frequently compared to previous `Poll` func.
As a result, this may lead to performance degradation rather than promotion.

Reply:
You are right, this is consistent with the benchmarking results.  While elapsed time is about the same for poll vs channel, the channel method has 2% higher user CPU time under stress.  Based on this, I'm inclined to go back to the original poll method.